File: mobile/lib/i18n.dart
/// Hand-rolled localization: no flutter_localizations/intl dependency.
///
/// Game terms (stages, phases, resources, item names) come verbatim from the
/// game's own localization files (data/i18n_glossary.json / the full APK
/// dictionary); app prose is translated by hand. English is the fallback for
/// any missing or empty entry.
///
/// The app persists and computes with INTERNAL keys ('Nascent', 'EARLY',
/// 'Legendary', ...); only display strings go through [tr] / [trStage] /
/// [trPhase], so saved inputs survive language switches.
library;

import 'dart:convert';

/// Supported languages, code -> native name (for the picker).
const langs = {
  'en': 'English',
  'ru': 'Русский',
  'de': 'Deutsch',
  'es': 'Español',
  'zh': '中文',
};

/// Current UI language code. The app does full-page rebuilds on change, so a
/// plain mutable top-level is enough (set from SharedPreferences before the
/// first build).
String currentLang = 'en';

/// Translate [s] into the current language; falls back to English ([s]
/// itself) when there is no (or an empty) translation.
String tr(String s) {
  if (currentLang == 'en') return s;
  final v = _t[s]?[currentLang];
  return (v == null || v.isEmpty) ? s : v;
}

/// Display name for an internal stage key ('Nascent' -> 'Пробуждение').
String trStage(String key) => tr(key);

const _phaseNames = {'EARLY': 'Early', 'MIDDLE': 'Middle', 'LATE': 'Late'};

/// Display name for an internal phase key ('EARLY' -> 'Early'/'Начальная').
String trPhase(String key) => tr(_phaseNames[key] ?? key);

/// Translations, loaded once from the shared data/i18n.json (the SAME
/// file the desktop app reads — single source of truth, so the two can
/// never drift). Populated by [loadTranslations] in main() before the
/// first frame; empty until then (tr falls back to English).
Map<String, Map<String, String>> _t = {};

/// Parse assets/data/i18n.json ({en: {lang: val}}) into [_t].
void loadTranslations(String jsonStr) {
  final raw = jsonDecode(jsonStr) as Map<String, dynamic>;
  _t = {
    for (final e in raw.entries)
      e.key: {
        for (final l in (e.value as Map<String, dynamic>).entries)
          l.key: l.value as String
      }
  };
}

/// Localize an engine-formatted duration string ("1D 12H 0M  (~1.2 yr)")
/// at display time. The engine's fmtDays stays canonical English — tests
/// and parity compare its exact output.
String trDuration(String s) {
  const suffixes = {
    'ru': ['д', 'ч', 'м', 'г'],
    'de': ['T', 'Std', 'Min', 'J'],
    'es': ['d', 'h', 'min', 'años'],
    'zh': ['天', '时', '分', '年'],
  };
  final suf = suffixes[currentLang];
  if (suf == null) return s;
  var out = s.replaceAllMapped(
      RegExp(r'(\d+)D (\d+)H (\d+)M'),
      (m) => currentLang == 'zh'
          ? '${m[1]}${suf[0]}${m[2]}${suf[1]}${m[3]}${suf[2]}'
          : '${m[1]}${suf[0]} ${m[2]}${suf[1]} ${m[3]}${suf[2]}');
  out = out.replaceAllMapped(
      RegExp(r'\(~([\d.]+) yr\)'),
      (m) => currentLang == 'zh' ? '(~${m[1]}${suf[3]})' : '(~${m[1]} ${suf[3]})');
  return out;
}

File: breakthrough_calc/i18n.py
"""Tiny gettext-style i18n for the desktop GUI.

tr(s) looks up the English source string in the current language's dict and
falls back to English (s itself). Translations are the SINGLE SOURCE OF
TRUTH shared with the mobile app: data/i18n.json ({en: {ru,de,es,zh}}),
synced to mobile/assets/data and pinned by tests. Desktop and mobile can
never drift because they read the same file — edit a string once, both
platforms get it.

reverse(s) maps a translated display string (any language) back to its
English source — used to turn combo display text back into internal keys
and to migrate legacy settings that stored localized display names.
"""

from __future__ import annotations

import json

from .data_io import resource_path

LANGS = {"en": "English", "ru": "Русский", "de": "Deutsch", "es": "Español", "zh": "中文"}


def _load_translations() -> dict:
    """data/i18n.json ({en: {lang: val}}) -> {lang: {en: val}} for tr()."""
    try:
        with open(resource_path("data", "i18n.json"), encoding="utf-8") as f:
            raw = json.load(f)
    except (OSError, ValueError):
        return {lang: {} for lang in LANGS if lang != "en"}
    out: dict = {lang: {} for lang in LANGS if lang != "en"}
    for en, row in raw.items():
        for lang, val in row.items():
            out.setdefault(lang, {})[en] = val
    return out

_current = "en"


def set_lang(code: str):
    global _current
    _current = code if code in LANGS else "en"


def get_lang() -> str:
    return _current


def tr(s: str) -> str:
    if _current == "en":
        return s
    return TRANSLATIONS.get(_current, {}).get(s, s)


def reverse(s: str) -> str:
    """Translated display string (any language) -> English source; identity
    for English/unknown strings."""
    return _REVERSE.get(s, s)


TRANSLATIONS = _load_translations()

# Translated display string -> English source, across all languages. The
# shared table holds both display labels ("Nascent Soul") and the raw
# internal keys the mobile app translates directly ("Nascent") — which can
# share a translation. reverse() is used to turn a combo's DISPLAYED text
# back into its English label, so on a collision the more specific (longer)
# key wins, keeping the display-label round-trip intact.
_REVERSE = {}
for _lang in TRANSLATIONS.values():
    for _en in sorted(_lang, key=lambda s: (-len(s), s)):
        _REVERSE.setdefault(_lang[_en], _en)


# Duration suffixes for display-time localization of engine fmt_days output.
# The engine string stays canonical English ("1D 12H 0M  (~1.2 yr)") — tests
# and cross-engine parity compare its exact output.
_DUR_SUFFIXES = {
    "ru": ("д", "ч", "м", "г"),
    "de": ("T", "Std", "Min", "J"),
    "es": ("d", "h", "min", "años"),
    "zh": ("天", "时", "分", "年"),
}

_DUR_RE = None
_YR_RE = None


def tr_duration(s: str) -> str:
    """Localize a fmt_days string at display time; identity for English."""
    global _DUR_RE, _YR_RE
    suf = _DUR_SUFFIXES.get(get_lang())
    if suf is None:
        return s
    import re
    if _DUR_RE is None:
        _DUR_RE = re.compile(r"(\d+)D (\d+)H (\d+)M")
        _YR_RE = re.compile(r"\(~([\d.]+) yr\)")
    joiner = "" if get_lang() == "zh" else " "
    s = _DUR_RE.sub(lambda m: f"{m[1]}{suf[0]}{joiner}{m[2]}{suf[1]}{joiner}{m[3]}{suf[2]}", s)
    s = _YR_RE.sub(lambda m: f"(~{m[1]}{joiner}{suf[3]})", s)
    return s
